src/Host/Broker/Impl/Sessions/Session.cs

```diff
@@ -75,105 +75,150 @@ public class Session {
+            if (useridentity != null && WindowsIdentity.GetCurrent().User != useridentity.User) {
+                uint error = NativeMethods.CredUIParseUserName(User.Name, usernameBldr, usernameBldr.Capacity, domainBldr, domainBldr.Capacity);
                 if (error != 0) {
                     _sessionLogger.LogError(Resources.Error_UserNameParse, User.Name, error);
                     throw new ArgumentException(string.Format(Resources.Error_UserNameParse, User.Name, error));
                 }
 
-                psi.Domain = domain.ToString();
-                psi.UserName = username.ToString();
-                psi.Password = password;
+                string domain = domainBldr.ToString();
+                string username = usernameBldr.ToString();
 
-                _sessionLogger.LogTrace(Resources.Trace_EnvironmentVariableCreationBegin, User.Name, profilePath);
-                // if broker and rhost are run as different users recreate user environment variables.
-                psi.EnvironmentVariables["USERNAME"] = username.ToString();
-                _sessionLogger.LogTrace(Resources.Trace_EnvironmentVariable, "USERNAME", psi.EnvironmentVariables["USERNAME"]);
+                Win32EnvironmentBlock eb = CreateEnvironmentBlockForUser(useridentity, username, profilePath);
 
-                psi.EnvironmentVariables["HOMEDRIVE"] = profilePath.Substring(0, 2);
-                _sessionLogger.LogTrace(Resources.Trace_EnvironmentVariable, "HOMEDRIVE", psi.EnvironmentVariables["HOMEDRIVE"]);
+                eb["R_HOME"] = shortHome.ToString();
+                eb["PATH"] = Interpreter.Info.BinPath + ";" + Environment.GetEnvironmentVariable("PATH");
 
-                psi.EnvironmentVariables["HOMEPATH"] = profilePath.Substring(2);
-                _sessionLogger.LogTrace(Resources.Trace_EnvironmentVariable, "HOMEPATH", psi.EnvironmentVariables["HOMEPATH"]);
+                _sessionLogger.LogInformation(Resources.Info_StartingRHost, Id, User.Name, rhostExePath, arguments);
+                Stream stderror;
+                int pid  = Win32Process.StartProcessAsUser(useridentity, rhostExePath, arguments, Path.GetDirectoryName(rhostExePath), eb, out stdin, out stdout, out stderror);
 
-                psi.EnvironmentVariables["USERPROFILE"] = $"{psi.EnvironmentVariables["HOMEDRIVE"]}{psi.EnvironmentVariables["HOMEPATH"]}";
-                _sessionLogger.LogTrace(Resources.Trace_EnvironmentVariable, "USERPROFILE", psi.EnvironmentVariables["USERPROFILE"]);
+                _process = Process.GetProcessById(pid);
+                _process.EnableRaisingEvents = true;
 
-                psi.EnvironmentVariables["APPDATA"] = $"{psi.EnvironmentVariables["USERPROFILE"]}\\AppData\\Roaming";
-                _sessionLogger.LogTrace(Resources.Trace_EnvironmentVariable, "APPDATA", psi.EnvironmentVariables["APPDATA"]);
+                _process.Exited += delegate {
```

Comment:
There is a race condition here now, since we start the process before subscribing the event. I'm not sure if `Process` does any magic to still raise the event in retrospect when a handler gets added, but most likely not. So this code might need to be factored into a helper method, and then you can call that helper both from here and from `if (HasExited)` below.